### PR TITLE
fenicsx: Add versions 0.4.0 and 0.4.1

### PR DESCRIPTION
**Pull request by bosmacs**
This pull request adds recently released versions of the fenicsx packages:
- fenics-basix
- fenics-dolfinx
- fenics-ffcx
- py-fenics-basix
- py-fenics-dolfinx
- py-fenics-ffcx

### CODE REVIEW

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/fenics-dolfinx/package.py:51]: FYI. You could actually use a for loop to set up these version-specific dependencies here and below, e.g.:

```suggestion
    for ver in ['0.1.0', '0.2.0', '0.3.0', '0.4.0', '0.4.1']:
        depends_on("py-fenics-ffcx@{0}".format(ver), type=("build", "run"), when="@{0}".format(ver))
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/fenics-dolfinx/package.py:51]: You could actually include `main` as a version for that matter.

**Review by tldahlgren — COMMENTED** [review]
Confirmed all of the sha256.  Just have a suggestion to tighten the code.

Let me know if you want to make the suggested changes in this PR.

**Review by tldahlgren — APPROVED** [review]

**Comment by garth-wells** [comment]
It might be helpful to hold off for a bit. We're testing some changes that will avoid making Python a dependency of  `fenics-dolfinx`, and tidying up some other dependencies. Some steps in this direction are at https://github.com/FEniCS/spack/pull/1/files.

I'd prefer to not include the older patch releases - we found a few bugs in the v0.4 release process that were fixes.

**Comment by tldahlgren** [comment]
> It might be helpful to hold off for a bit. We're testing some changes that will avoid making Python a dependency of `fenics-dolfinx`, and tidying up some other dependencies. Some steps in this direction are at https://github.com/FEniCS/spack/pull/1/files.
> 
> I'd prefer to not include the older patch releases - we found a few bugs in the v0.4 release process that were fixes.

You might be able to switch this to a Draft PR until you're ready.  Regardless, ping me when you're ready.

File: var/spack/repos/builtin/packages/fenics-dolfinx/package.py
Comment (bosmacs) [reply var/spack/repos/builtin/packages/fenics-dolfinx/package.py:51]: This is a good suggestion. If other revisions are necessary I will include this as well.

**Comment by bosmacs** [comment]
@garth-wells Is there an expectation of when those changes will land?

I'm content to hold off on this pull request if the fenics maintainers plan to have an update soon. I made these updates initially for myself to be in sync with known releases, and as a new spack user thought it might be useful to others.

**Comment by garth-wells** [comment]
@bosmacs change will hopefully land very soon. PR is open at https://github.com/spack/spack/pull/30661. Needs some small tweaks still.

**Comment by garth-wells** [comment]
Updates in #30661.

**Pull request closed without merging**
